utils/keys_test.go

```diff
@@ -252,17 +274,70 @@ func TestImportNoPath(t *testing.T) {
 	b := &pem.Block{
 		Headers: make(map[string]string),
 	}
-	b.Bytes = make([]byte, 1000)
-	rand.Read(b.Bytes)
+	from, _ := os.OpenFile("../fixtures/secure.example.com.key", os.O_RDONLY, notary.PrivKeyPerms)
+	defer from.Close()
+	b.Bytes, _ = ioutil.ReadAll(from)
 
-	bBytes := pem.EncodeToMemory(b)
+	in := bytes.NewBuffer(b.Bytes)
+
+	err := ImportKeys(in, []Importer{s}, "root", "", passphraseRetriever)
+	require.NoError(t, err)
+
+	for key := range s.data {
+		// no path but role included should work
+		require.Equal(t, key, filepath.Join(notary.RootKeysSubdir, "12ba0e0a8e05e177bc2c3489bdb6d28836879469f078e68a4812fc8a2d521497"))
+	}
+
+	err = ImportKeys(in, []Importer{s}, "", "", passphraseRetriever)
+	require.NoError(t, err)
+
+	require.Len(t, s.data, 1) // no path and no role should not work
+}
+
+func TestNonRootPathInference(t *testing.T) {
+	s := NewTestImportStore()
+
+	b := &pem.Block{
+		Headers: make(map[string]string),
+	}
+	from, _ := os.OpenFile("../fixtures/secure.example.com.key", os.O_RDONLY, notary.PrivKeyPerms)
+	defer from.Close()
+	b.Bytes, _ = ioutil.ReadAll(from)
+
+	in := bytes.NewBuffer(b.Bytes)
+
+	err := ImportKeys(in, []Importer{s}, data.CanonicalSnapshotRole, "somegun", passphraseRetriever)
+	require.NoError(t, err)
+
+	for key := range s.data {
+		// no path but role included should work and 12ba0e0a8e05e177bc2c3489bdb6d28836879469f078e68a4812fc8a2d521497 is the key ID of the fixture
+		require.Equal(t, key, filepath.Join(notary.NonRootKeysSubdir, "somegun", "12ba0e0a8e05e177bc2c3489bdb6d28836879469f078e68a4812fc8a2d521497"))
+	}
+}
+
+func TestBlockHeaderPrecedence(t *testing.T) {
+	s := NewTestImportStore()
+
+	b := &pem.Block{
+		Headers: make(map[string]string),
+	}
+	from, _ := os.OpenFile("../fixtures/precedence.example.com.key", os.O_RDONLY, notary.PrivKeyPerms)
+	defer from.Close()
+	b.Bytes, _ = ioutil.ReadAll(from)
 
-	in := bytes.NewBuffer(bBytes)
+	in := bytes.NewBuffer(b.Bytes)
 
-	err := ImportKeys(in, []Importer{s})
+	err := ImportKeys(in, []Importer{s}, "somerole", "somegun", passphraseRetriever)
 	require.NoError(t, err)
 
-	require.Len(t, s.data, 0)
+	for key := range s.data {
+		// block header role= root should take precedence over command line flag
```

Comment:
👍  Awesome test, thank you! 

Could we also add one for falling back on the delegation role if no role is given and no role is provided?  (also maybe one for overriding the gun if the path specifies a different gun?)